types/react-query/index.d.ts

```diff
@@ -1,142 +1,370 @@
+    config?: QueryOptions<TResult>,
+): PaginatedQueryResult<TResult>;
+
+export function usePaginatedQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables = []>({
+    queryKey,
+    variables,
+    queryFn,
+    config,
+}: {
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined);
+    variables?: TVariables;
+    queryFn: QueryFunctionWithVariables<TResult, TKey, TVariables>;
+    config?: QueryOptions<TResult>;
+}): PaginatedQueryResult<TResult>;
+
+// useInfiniteQuery
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, TKey, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export type QueryFunction<TResult, TVariables extends object> = (variables: TVariables) => Promise<TResult>;
+export function useInfiniteQuery<TResult, TKey extends string, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, [TKey], TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export interface QueryOptions<TResult> {
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends string, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, [TKey], TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<
+    TResult,
+    TKey extends AnyQueryKey,
+    TMoreVariable,
+    TVariables extends AnyVariables = []
+>({
+    queryKey,
+    variables,
+    queryFn,
+    config,
+}: {
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined);
+    variables?: TVariables;
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>;
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>;
+}): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export type QueryKeyPart = string | object | boolean | number | null | readonly QueryKeyPart[] | null | undefined;
+export type AnyQueryKey =
```

Comment:
any reason for not using tuple type? `[string, ... QueryKeyPart[]]`?

Reply:
>  it can be any length of any items that are serializable
https://github.com/tannerlinsley/react-query/issues/257#issuecomment-600441958

Reply:
This was a tradeoff. The way the key constituents are to the query function makes it hard to reconstruct the signature of the query function. This is how a query function is being called: `queryFn(...queryKey, ...queryVariables)` or in case of infinite scroll `queryFn(...queryKey, ...queryVariables, moreVariable)`. In this case if the type of the key is inferred as an array, you may not have a reliable signature of the query function. It will be essentially something like this `(...keysOrVariablesOrMore: (TKey[number] | TVariables[number] | TMoreVariable | undefined)[]) => Promise<TResult>`.
So I chose to have a reasonable limitation to always use tuples up to some size. We can increase the maximum size at any time.

I actually experimented just now how to enforce TS to infer tuples instead of arrays. It seems with a few tricks it is possible to have any tuple size. I'll update the code shortly.

Reply:
But in the every definition of query function you're using `List.Concat` to create a type of args that will be passed to function:

``` type QueryFunctionWithVariables<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables> = (
    ...key: _.List.Concat<TKey, TVariables>
) => Promise<TResult>;
```

Unless I'm not seeing something here, this will also work if `TKey` and `TVars` are constrained to open-ended tuples (`[string, ...any[]]`)

Reply:
If the key is open-ended, then you can say in which parameter position variables or fetch more variable will land. The latest code has commented out example where it fails. `_.List.Concat` cannot concat tuples if the left is open-ended: `_.List.Concat<[1, ...2[]], [3, 4]> = [3, 4]` when actually should be `[1, 2 | 3, 2| 4, ...2[]]`.
I found a way to do it with `typescript-tuple` but it needs to be approved first (https://github.com/microsoft/types-publisher/pull/758).